File: crates/swarm-p2p/src/punch.rs
//! Simultaneous UDP hole punching (`docs/PROTOCOL.md`'s connection-
//! establishment step 4): both sides blast `PUNCH_MAGIC` at every candidate
//! address, LAN first, while listening on the same socket for the same
//! magic coming back the other way.
//!
//! Deliberately signaling-agnostic: [`punch`] only proves *this device* can
//! receive from one of `candidates` — it doesn't know about
//! `swarm_stun_client::SignalingClient` or send `Punched` itself. Mutual
//! confirmation (so neither side switches to the punched path while the
//! other is still waiting, per the protocol doc) is the caller's job,
//! sitting one layer up where both this crate and the signaling client are
//! already in scope — matches how `apps/server`'s `ServerCore` is the thing
//! that wires `swarm-stun-client` and `swarm-p2p` together, not either
//! crate depending on the other.

use std::net::SocketAddr;
use std::time::Duration;
use swarm_core::signal::PUNCH_MAGIC;
use tokio::net::UdpSocket;
use tokio::time::Instant;

pub const MAX_ATTEMPTS: u32 = 20;
pub const ATTEMPT_INTERVAL: Duration = Duration::from_millis(200);

#[derive(Debug, thiserror::Error)]
pub enum PunchError {
    #[error("no response from any of {0} candidate(s) after {1} attempts")]
    NoResponse(usize, u32),
    #[error("socket error while punching: {0}")]
    Network(#[from] std::io::Error),
}

/// Tries every address in `candidates` (in the given order, each round —
/// ordering LAN-first is the caller's job) every [`ATTEMPT_INTERVAL`], up to
/// [`MAX_ATTEMPTS`] times, while listening on `socket` for the same magic
/// coming back. Returns the source address of the first valid magic packet
/// received from a listed candidate, having already sent one more magic
/// packet back to that exact address on the way out — the scheduled blast
/// for this round may have gone out to a different candidate that didn't
/// pan out, so this makes sure a packet flows back down the *confirmed*
/// path immediately rather than waiting for the next round.
pub async fn punch(socket: &UdpSocket, candidates: &[SocketAddr]) -> Result<SocketAddr, PunchError> {
    if candidates.is_empty() {
        return Err(PunchError::NoResponse(0, 0));
    }
    let mut buf = [0u8; PUNCH_MAGIC.len()];
    for _attempt in 0..MAX_ATTEMPTS {
        for &candidate in candidates {
            socket.send_to(PUNCH_MAGIC, candidate).await?;
        }
        let deadline = Instant::now() + ATTEMPT_INTERVAL;
        loop {
            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                break;
            }
            match tokio::time::timeout(remaining, socket.recv_from(&mut buf)).await {
                Ok(Ok((len, from))) if &buf[..len] == PUNCH_MAGIC && candidates.contains(&from) => {
                    let _ = socket.send_to(PUNCH_MAGIC, from).await; // best-effort immediate reply
                    return Ok(from);
                }
                Ok(Ok(_)) => continue,       // stray/malformed datagram — keep waiting out this round
                Ok(Err(e)) => return Err(PunchError::Network(e)),
                Err(_) => break,             // this round's window elapsed — try the next round
            }
        }
    }
    Err(PunchError::NoResponse(candidates.len(), MAX_ATTEMPTS))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn simultaneous_punch_succeeds_both_directions() {
        let a = UdpSocket::bind("127.0.0.1:0").await.unwrap();
        let b = UdpSocket::bind("127.0.0.1:0").await.unwrap();
        let a_addr = a.local_addr().unwrap();
        let b_addr = b.local_addr().unwrap();

        let (a_candidates, b_candidates) = ([b_addr], [a_addr]);
        let (a_result, b_result) = tokio::join!(punch(&a, &a_candidates), punch(&b, &b_candidates));
        assert_eq!(a_result.unwrap(), b_addr);
        assert_eq!(b_result.unwrap(), a_addr);
    }

    #[tokio::test]
    async fn tries_every_candidate_each_round_not_just_the_first() {
        let listener = UdpSocket::bind("127.0.0.1:0").await.unwrap();
        let listener_addr = listener.local_addr().unwrap();
        let dead_candidate: SocketAddr = "127.0.0.1:1".parse().unwrap(); // nothing responds here

        let puncher = UdpSocket::bind("127.0.0.1:0").await.unwrap();

        let listener_task = tokio::spawn(async move {
            let mut buf = [0u8; 64];
            let (len, from) = listener.recv_from(&mut buf).await.unwrap();
            assert_eq!(&buf[..len], PUNCH_MAGIC);
            listener.send_to(PUNCH_MAGIC, from).await.unwrap();
        });

        let result = punch(&puncher, &[dead_candidate, listener_addr]).await.unwrap();
        assert_eq!(result, listener_addr);
        listener_task.await.unwrap();
    }

    #[tokio::test]
    async fn no_response_is_a_no_response_error() {
        let socket = UdpSocket::bind("127.0.0.1:0").await.unwrap();
        let nobody: SocketAddr = "127.0.0.1:1".parse().unwrap();
        let err = punch(&socket, &[nobody]).await.unwrap_err();
        assert!(matches!(err, PunchError::NoResponse(1, MAX_ATTEMPTS)));
    }

    #[tokio::test]
    async fn magic_from_an_unlisted_address_is_ignored() {
        let puncher = UdpSocket::bind("127.0.0.1:0").await.unwrap();
        let intended_candidate: SocketAddr = "127.0.0.1:1".parse().unwrap(); // never actually responds

        // A third party sends valid magic, but wasn't in the candidate list.
        let stranger = UdpSocket::bind("127.0.0.1:0").await.unwrap();
        let puncher_addr = puncher.local_addr().unwrap();
        tokio::spawn(async move {
            stranger.send_to(PUNCH_MAGIC, puncher_addr).await.unwrap();
        });

        let err = punch(&puncher, &[intended_candidate]).await.unwrap_err();
        assert!(matches!(err, PunchError::NoResponse(..)));
    }
}

File: clients/tv-android/core/src/main/kotlin/app/swarm/tv/core/transport/Punch.kt
/**
 * Simultaneous UDP hole punching, mirroring `swarm_p2p::punch::punch`
 * (Rust): both sides blast `PUNCH_MAGIC` at every candidate address, LAN
 * first, while listening on the same socket for the same magic coming back
 * the other way.
 *
 * Deliberately signaling-agnostic, same as the Rust version: [punch] only
 * proves *this device* can receive from one of `candidates` — it knows
 * nothing about `SignalingClient`. Mutual confirmation (so neither side
 * switches to the punched path while the other is still waiting, per
 * `docs/PROTOCOL.md`) is whatever orchestrates this one layer up.
 */
package app.swarm.tv.core.transport

import app.swarm.tv.core.signal.PUNCH_MAGIC
import java.io.IOException
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetSocketAddress
import java.net.SocketTimeoutException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

const val PUNCH_MAX_ATTEMPTS = 20
const val PUNCH_ATTEMPT_INTERVAL_MS = 200L

sealed class PunchError(message: String) : Exception(message) {
    class NoResponse(candidateCount: Int, attempts: Int) :
        PunchError("no response from any of $candidateCount candidate(s) after $attempts attempts")
    class Network(cause: Throwable) : PunchError("socket error while punching: ${cause.message}")
}

/**
 * Tries every address in [candidates] (in the given order, each round —
 * ordering LAN-first is the caller's job) every [PUNCH_ATTEMPT_INTERVAL_MS],
 * up to [PUNCH_MAX_ATTEMPTS] times, while listening on [socket] for the same
 * magic coming back. Returns the source address of the first valid magic
 * packet received from a listed candidate, having already sent one more
 * magic packet back to that exact address on the way out — the scheduled
 * blast for this round may have gone out to a different candidate that
 * didn't pan out, so this makes sure a packet flows back down the
 * *confirmed* path immediately rather than waiting for the next round.
 */
suspend fun punch(socket: DatagramSocket, candidates: List<InetSocketAddress>): InetSocketAddress =
    withContext(Dispatchers.IO) {
        if (candidates.isEmpty()) throw PunchError.NoResponse(0, 0)
        val originalTimeout = socket.soTimeout
        try {
            repeat(PUNCH_MAX_ATTEMPTS) {
                for (candidate in candidates) sendMagic(socket, candidate)
                receiveMagicWithin(socket, candidates, PUNCH_ATTEMPT_INTERVAL_MS)?.let { from ->
                    sendMagic(socket, from) // best-effort immediate reply
                    return@withContext from
                }
            }
            throw PunchError.NoResponse(candidates.size, PUNCH_MAX_ATTEMPTS)
        } finally {
            socket.soTimeout = originalTimeout
        }
    }

private fun sendMagic(socket: DatagramSocket, to: InetSocketAddress) {
    try {
        socket.send(DatagramPacket(PUNCH_MAGIC, PUNCH_MAGIC.size, to))
    } catch (e: IOException) {
        throw PunchError.Network(e)
    }
}

/**
 * Listens for up to [windowMs], returning the source address of the first
 * valid magic packet from a listed candidate, or null if the window elapses
 * first. A stray or malformed datagram doesn't end the wait, same as the
 * Rust version — it just keeps listening out the rest of the window.
 */
private fun receiveMagicWithin(socket: DatagramSocket, candidates: List<InetSocketAddress>, windowMs: Long): InetSocketAddress? {
    val buffer = ByteArray(PUNCH_MAGIC.size)
    val deadline = System.nanoTime() + windowMs * 1_000_000
    while (true) {
        val remainingMs = (deadline - System.nanoTime()) / 1_000_000
        if (remainingMs <= 0) return null
        socket.soTimeout = remainingMs.toInt().coerceAtLeast(1)
        val packet = DatagramPacket(buffer, buffer.size)
        try {
            socket.receive(packet)
        } catch (e: SocketTimeoutException) {
            return null
        } catch (e: IOException) {
            throw PunchError.Network(e)
        }
        val from = packet.socketAddress as InetSocketAddress
        val isValidMagic = packet.length == PUNCH_MAGIC.size && buffer.copyOf(packet.length).contentEquals(PUNCH_MAGIC)
        if (isValidMagic && candidates.contains(from)) return from
    }
}
